src/profiler.ts

```diff
@@ -13,57 +13,222 @@
+ */
+export interface RequestProfile {
+  name?: string;
+  profileType?: string;
+  duration?: any;
+  profileBytes?: string;
+  labels?: {instance?: string; zone?: string};
 }
 
-// TODO: finish implementing Profiler.
-// TODO: add stop() method to stop profiling.
-export class Profiler {
-  config: ProfilerConfig;
-  logger: Logger;
-  service: Service;
+/**
+ * Converts a profile to a compressed, base64 encoded string.
+ *
+ * @param p - profile to be converted to string.
+ */
+async function profileBytes(p: perftools.profiles.IProfile): Promise<string> {
+  const pwriter = perftools.profiles.Profile.encode(p);
+  const buffer = new Buffer(pwriter.finish());
+  const gzBuf = await gzip(buffer);
+  return gzBuf.toString('base64');
+}
+
+/**
+ * Polls Stackdriver Profiler server for instructions on behalf of a task and
+ * collects and uploads profiles as requested
+ */
+export class Profiler extends common.ServiceObject {
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private profileTypes: string[];
+  private continueProfiling: boolean;
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
-    this.config = config;
+    config = common.util.normalizeArguments(null, config);
+    const serviceConfig = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(serviceConfig, config), baseUrl: '/'});
 
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
+    this.config = config;
 
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(WALL_TYPE);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    while (true) {
```

Comment:
As discussed in person, we need to get rid of the this infinite loop as it will keep the application alive forever.

Instead we should pick a minimum interval (say `minInterval`) that we want to allow between profiles and setTimeout/delay for `minInterval - time taken by createProfile`. The server controls the max delay between profiles, but the agent should have a notion of the minimum delay.

Secondly we need a comment before the call to createProfile to indicate that it uses a hanging `GET`. Looking at the code it seems that it has "only" the latency of a normal http GET, whereas in reality it a lot longer.